duckduckgo/Android
# FIX bug that makes clean data not to work on app exit when VPN enabled

Task/Issue URL: https://app.asana.com/0/414730916066338/1201439663288660/f

### Description
With AppTP we introduced a bug that manifests itself when the AppTP is ENABLED

_Repro steps_
* install app and enable appTP
* open couple tabs
* go to Settings → Automatic clear
  * set Tabs and Data
* Close app and re-open
* Expected
  * data is clear
* Actual
  * data is not clear

The problem is in the logic in `AutomaticDataClearer`. The class keeps some state, in particular the `isFreshAppLaunch` field. That it is assumed to be set back to `true` when the app is closed and re-open.
This is becasue the `AutomaticDataClearer` assumes/expects that Android will kill the app process when the user swipes the app closed, however
* this is true when the VPN service is OFF
* but it is NOT true when the VPN service is ON
  * in that case, swiping the app closed will NOT kill the app process
  * `AutomaticDataClearer` instance is not cleaned and so the `isFreshAppLaunch` will remain set to `false` not clearing the data


_Solution_
The only reliable way to detect whether the app is closed with and without VPN enabled is to track the app activities states. When the app does not have any activity in CREATED state, we consider the app is closed.
* a new class `BrowserApplicationStateInfo` that performs that tracking
* and calls `BrowserLifecycleObserver` observer callbacks
  * `onOpen`, `onForeground`, `onBackground`, `onClosed` and `onExit`
  * see javadocs in `BrowserLifecycleObserver` class to understahdn when they are called

_Why the new gradle module browser-api?_
* I considered that `BrowserLifecycleObserver` would also be handy in other scenarios as oppose to rely on `LifecycleObserver`
* as `BrowserLifecycleObserver` may get used from other gradle modules I already advanced that and created the `browser-api` module
  * `browser-api` module will be home for any public API from the browser app
  * `browser-api` was already planned to be added under [this project](https://app.asana.com/0/72649045549333/1199214002677571/f). This PR just advances that a bit

### Steps to test this PR

_Clear tabs and data_
1. install from this branch
1. go to settings -> automatic clear and configure to clear "tabs and data" on "app exit"
1. open couple tabs
1. in one of the tabs, search for eg. "hola" and click in one of the links
1. open another tab and visit https://httpbin.org/basic-auth/u/p
  * login with user: u and password: p
1. home the app and reopen
1. VERIFY clear didn't happen and tabs are still there
1. swipe close the app and reopen
1. VERIFY tabs are not there
1. search for "hola" and verify that the previously opened link is not marked as visited
1. visit https://httpbin.org/basic-auth/u/p
1. VERIFY you will need to input the login credentials again

_Clear tabs_
1. install from this branch
1. go to settings -> automatic clear and configure to clear "tabs" on "app exit"
1. open couple tabs
1. in one of the tabs, search for eg. "hola" and click in one of the links
1. open another tab and visit https://httpbin.org/basic-auth/u/p
  * login with user: u and password: p
1. home the app and reopen
1. VERIFY clear didn't happen and tabs are still there
1. swipe close the app and reopen
1. VERIFY tabs are not there
1. search for "hola" and verify that the previously opened link is not marked as visited
1. visit https://httpbin.org/basic-auth/u/p
1. VERIFY you will NOT need to input the login credentials again
